metadata-io/src/main/java/com/linkedin/metadata/trace/KafkaTraceReader.java

```diff
@@ -0,0 +1,460 @@
+      int pollDurationMillis,
+      int pollMaxAttempts,
+      ExecutorService executorService,
+      long timeoutSeconds) {
+    this.adminClient = adminClient;
+    this.consumerSupplier = consumerSupplier;
+    this.pollDurationMs = pollDurationMillis;
+    this.pollMaxAttempts = pollMaxAttempts;
+    this.executorService = executorService;
+    this.timeoutSeconds = timeoutSeconds;
+  }
+
+  @Nonnull
+  protected abstract String getTopicName();
+
+  @Nullable
+  protected abstract String getConsumerGroupId();
+
+  public abstract Optional<T> read(@Nullable GenericRecord genericRecord);
+
+  protected abstract Optional<Pair<ConsumerRecord<String, GenericRecord>, SystemMetadata>>
+      matchConsumerRecord(
+          ConsumerRecord<String, GenericRecord> consumerRecord, String traceId, String aspectName);
+
+  /**
+   * Determines the write status of a trace by comparing consumer offset with message offset.
+   *
+   * @return PENDING if the message exists but hasn't been consumed yet, UNKNOWN if no consumer
+   *     offset exists, ERROR in other cases
+   */
+  public Map<Urn, Map<String, TraceStorageStatus>> tracePendingStatuses(
+      Map<Urn, List<String>> urnAspectPairs, String traceId, Long traceTimestampMillis) {
+    return tracePendingStatuses(urnAspectPairs, traceId, traceTimestampMillis, false);
+  }
+
+  public Map<Urn, Map<String, TraceStorageStatus>> tracePendingStatuses(
+      Map<Urn, List<String>> urnAspectPairs,
+      String traceId,
+      Long traceTimestampMillis,
+      boolean skipCache) {
+
+    List<CompletableFuture<Map.Entry<Urn, Map<String, TraceStorageStatus>>>> futures =
+        urnAspectPairs.entrySet().stream()
+            .map(
+                entry ->
+                    CompletableFuture.supplyAsync(
+                        () -> {
+                          try {
+                            Map<String, TraceStorageStatus> result =
+                                tracePendingStatuses(
+                                    entry.getKey(),
+                                    entry.getValue(),
+                                    traceId,
+                                    traceTimestampMillis,
+                                    skipCache);
+                            return Map.entry(entry.getKey(), result);
+                          } catch (Exception e) {
+                            log.error(
+                                "Error processing trace status for URN: {}", entry.getKey(), e);
+                            return Map.entry(
+                                entry.getKey(), Collections.<String, TraceStorageStatus>emptyMap());
+                          }
+                        },
+                        executorService))
+            .collect(Collectors.toList());
+
+    try {
+      List<Map.Entry<Urn, Map<String, TraceStorageStatus>>> results =
+          CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
```

Comment:
very minor nit: we already have the required size from futures so can avoid some array resizing costs by setting new CompletableFuture[futures.size()]), compiler might already do that though